src/emulator/modules/SceTouch/SceTouch.cpp

```diff
@@ -24,14 +24,95 @@
 
 // TODO Move elsewhere.
 static uint64_t timestamp;
+SDL_TouchFingerEvent finger_buffer[8];
+uint8_t finger_count = 0;
+long int finger_id_ref;
+auto touchscreen_port = SCE_TOUCH_PORT_FRONT;
+
+SceTouchData recover_touch_events() {
+    SceTouchData touch_data;
+    touch_data.timeStamp = timestamp;
+
+    for (int i = 0; i < finger_count; i++) {
+        touch_data.report[i].id = static_cast<uint8_t>(finger_buffer[i].fingerId - finger_id_ref);
+        touch_data.report[i].force = static_cast<uint8_t>(1 + finger_buffer[i].pressure * 127);
+        touch_data.report[i].x = static_cast<uint16_t>(finger_buffer[i].x * 1920);
+
+        if (touchscreen_port == SCE_TOUCH_PORT_FRONT) {
+            touch_data.report[i].y = static_cast<uint16_t>(finger_buffer[i].y * 1088);
+        } else {
+            touch_data.report[i].y = static_cast<uint16_t>(108 + finger_buffer[i].y * (889-108));
+        }
+    }
+
+    touch_data.reportNum = finger_count;
+
+    return touch_data;
+}
+
+int handle_touch_event(SDL_TouchFingerEvent finger) {
+    switch (finger.type) {
+    case SDL_FINGERDOWN: {
+        if (finger_count == 0) {
+            finger_id_ref = finger.fingerId;
+        }
+
+        finger_buffer[finger_count] = finger;
+        finger_count++;
+
+        break;
+    }
+    case SDL_FINGERUP: {
+        int finger_index = -1;
+        for (int i = 0; i < finger_count; i++) {
+            if (finger.fingerId == finger_buffer[i].fingerId) {
+                finger_index = i;
+            }
+            if (finger_index > -1) {
+                finger_buffer[i] = finger_buffer[i + 1];
+            }
+        }
+        finger_count--;
+        break;
+    }
+
+    case SDL_FINGERMOTION: {
+        for (int i = 0; i < finger_count; i++) {
+            if (finger.fingerId == finger_buffer[i].fingerId) {
+                finger_buffer[i] = finger;
+            }
+        }
+        break;
+    }
+    }
+
+    return 0;
+}
+
+bool registered_touch() {
```

Comment:
ditto

Reply:
whats ditto aprat from a pokemon?

Reply:
like upper, said before. ditto clones other in some way right?

Reply:
I think he means make it `static`.